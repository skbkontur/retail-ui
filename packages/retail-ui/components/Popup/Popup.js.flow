/* @flow */
export type PopupPosition =
  | 'top left'
  | 'top center'
  | 'top right'
  | 'right top'
  | 'right middle'
  | 'right bottom'
  | 'bottom left'
  | 'bottom center'
  | 'bottom right'
  | 'left top'
  | 'left middle'
  | 'left bottom';

declare export var PopupPositions: PopupPosition[];

export type PopupHandlerProps = {|
  onMouseEnter?: (event: SyntheticMouseEvent<HTMLElement> | MouseEvent) => void,
  onMouseLeave?: (event: SyntheticMouseEvent<HTMLElement> | MouseEvent) => void,
  onClick?: (event: SyntheticMouseEvent<HTMLElement> | MouseEvent) => void,
  onFocus?: (event: SyntheticFocusEvent<HTMLElement> | FocusEvent) => void,
  onBlur?: (event: SyntheticFocusEvent<HTMLElement> | FocusEvent) => void,
  onOpen?: () => void,
|};

export type PopupProps = {|
  ...$Exact<PopupHandlerProps>,
  anchorElement: React$Node | HTMLElement,
  backgroundColor?: string,
  borderColor?: string,
  children: React$Node | (() => React$Node),
  hasPin: boolean,
  hasShadow: boolean,
  disableAnimations: boolean,
  margin: number,
  maxWidth: number | string,
  opened: boolean,
  pinOffset: number,
  pinSize: number,
  popupOffset: number,
  positions: string[],
  onCloseRequest?: () => void,
|};

type PopupLocation = {|
  coordinates: {|
    left: number,
    top: number,
  |},
  position: string,
|};

export type PopupState = {|
  location: ?PopupLocation,
|};

declare class Popup extends React$Component<PopupProps, PopupState> {}

export default Popup;
