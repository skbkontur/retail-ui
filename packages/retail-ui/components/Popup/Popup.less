@import '../variables.less';

@show-transtion: transform 0.18s cubic-bezier(0.22, 0.61, 0.36, 1),
  opacity 0.18s cubic-bezier(0.22, 0.61, 0.36, 1);
@hide-transition: opacity 0.15s ease-out;

@value popupBackground: @bg-default;
@value popupBorderColor: @popup-border-color;

:local {
  .popup {
    position: absolute;
    min-width: 18px;
    border-radius: @popup-border-radius;
    border: @popup-border @popup-border-color;

    &-ignore-hover {
      pointer-events: none;
    }
  }

  .content {
    border-radius: @popup-border-radius;
    overflow: hidden;
  }

  .contentInner {
    background: @bg-default;
  }

  :global(.rt-ie-any) {
    .shadow {
      box-shadow: @popup-box-shadow;
    }
  }

  :global(.rt-ie8) {
    .popup {
      border: 1px solid @tooltip-border;
    }
  }

  .shadow {
    filter: @popup-drop-shadow;
    -webkit-filter: @popup-drop-shadow;
  }

  .transition {
    &-enter {
      opacity: 0.01;

      &-top {
        transform: translateY(10px);
      }

      &-bottom {
        transform: translateY(-10px);
      }

      &-left {
        transform: translateX(10px);
      }

      &-right {
        transform: translateX(-10px);
      }
    }

    &-enter-active {
      transition: @show-transtion;
      opacity: 1;
      transform: translate(0, 0);
    }

    &-exit {
      opacity: 0.01;
      transition: @hide-transition;
      transform: translate(0, 0);
    }
  }
}
