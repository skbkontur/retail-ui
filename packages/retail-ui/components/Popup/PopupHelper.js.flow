/* @flow */
export type Rect = {|
  top: number,
  left: number,
  width: number,
  height: number,
|};

export type Offset = {|
  top: number,
  left: number,
|};

export type PositionObject = {|
  direction: string,
  align: string,
|};

declare function getPositionObject(position: string): PositionObject;
declare function getElementAbsoluteRect(element: HTMLElement): Rect;
declare function isAbsoluteRectFullyVisible(absoluteRect: Rect): boolean;

declare var _default: {|
  getPositionObject: typeof getPositionObject,
  getElementAbsoluteRect: typeof getElementAbsoluteRect,
  isAbsoluteRectFullyVisible: typeof isAbsoluteRectFullyVisible,
|};

export default _default;
