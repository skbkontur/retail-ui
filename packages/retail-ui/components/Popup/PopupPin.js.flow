/* @flow */

type Props = {|
  backgroundColor: string,
  borderColor: string,
  borderWidth: number,
  offset: number,
  popupElement: ?HTMLElement,
  popupPosition: string,
  size: number,
|};

declare class PopupPin extends React$Component<Props> {}

export default PopupPin;
