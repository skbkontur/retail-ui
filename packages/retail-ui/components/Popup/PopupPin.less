:local {
  .wrapper {
    overflow: hidden;
    position: absolute;
    pointer-events: none;
  }
}
