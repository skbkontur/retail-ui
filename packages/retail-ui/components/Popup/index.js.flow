/* @flow */
export { default, PopupProps, PopupPosition, PopupPositions } from './Popup';
