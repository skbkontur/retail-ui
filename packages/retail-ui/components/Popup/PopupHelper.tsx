import { PopupPosition } from './Popup';

export interface Rect {
  top: number;
  left: number;
  width: number;
  height: number;
}

export interface Offset {
  top: number;
  left: number;
}

export interface PositionObject {
  direction: string;
  align: string;
}

function getPositionObject(position: string): PositionObject {
  const x = position.split(' ');

  return {
    direction: x[0],
    align: x[1],
  };
}

function getElementAbsoluteRect(element: HTMLElement): Rect {
  const rect = _getElementRelativeRect(element);
  return convertRectToAbsolute(rect);
}

function isAbsoluteRectFullyVisible(coordinates: Offset, popupRect: Rect): boolean {
  const windowRelativeRect = _getWindowRelativeRect();
  const windowAbsoluteRect = convertRectToAbsolute(windowRelativeRect);
  const absoluteRect = {
    top: coordinates.top,
    left: coordinates.left,
    height: popupRect.height,
    width: popupRect.width,
  };

  return _rectContainsRect(windowAbsoluteRect, absoluteRect);
}

function canBecomeFullyVisible(positionName: PopupPosition, coordinates: Offset) {
  const position = getPositionObject(positionName);

  if (position.direction === 'top') {
    const availableScrollDistances = convertRectToAbsolute(_getWindowRelativeRect());
    return coordinates.top + availableScrollDistances.top >= 0;
  }

  if (position.direction === 'left') {
    const availableScrollDistances = convertRectToAbsolute(_getWindowRelativeRect());
    return coordinates.left + availableScrollDistances.left >= 0;
  }

  // NOTE: for bottom/right cases browser will always expand document size
  return true;
}

function _getElementRelativeRect(element: HTMLElement) {
  const rect = element.getBoundingClientRect();

  return {
    top: rect.top,
    left: rect.left,
    width: rect.right - rect.left,
    height: rect.bottom - rect.top,
  };
}

function _getWindowRelativeRect(): Rect {
  return {
    top: 0,
    left: 0,
    width: _getViewProperty(x => x.clientWidth) || window.innerWidth,
    height: _getViewProperty(x => x.clientHeight) || window.innerHeight,
  };
}

function convertRectToAbsolute(rect: Rect): Rect {
  const offset = _getAbsoluteOffset();

  return {
    top: rect.top + offset.top,
    left: rect.left + offset.left,
    width: rect.width,
    height: rect.height,
  };
}

function _getAbsoluteOffset(): Offset {
  const scrollTop = window.pageYOffset || _getViewProperty(x => x.scrollTop);
  const scrollLeft = window.pageXOffset || _getViewProperty(x => x.scrollLeft);

  const clientTop = _getViewProperty(x => x.clientTop);
  const clientLeft = _getViewProperty(x => x.clientLeft);

  const top = scrollTop - clientTop;
  const left = scrollLeft - clientLeft;

  return {
    top,
    left,
  };
}

function _rectContainsRect(outerRect: Rect, innerRect: Rect): boolean {
  return (
    innerRect.top > outerRect.top &&
    innerRect.top + innerRect.height < outerRect.top + outerRect.height &&
    innerRect.left > outerRect.left &&
    innerRect.left + innerRect.width < outerRect.left + outerRect.width
  );
}

function _getViewProperty(getProperty: (e: HTMLElement) => number): number {
  const views = [document.documentElement, document.body];
  return views.map(x => x && getProperty(x)).find(Boolean) || 0;
}

export default {
  getPositionObject,
  getElementAbsoluteRect,
  isFullyVisible: isAbsoluteRectFullyVisible,
  canBecomeFullyVisible,
};
