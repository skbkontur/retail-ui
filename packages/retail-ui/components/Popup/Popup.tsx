import cn from 'classnames';
import * as React from 'react';
import { findDOMNode } from 'react-dom';
import * as PropTypes from 'prop-types';
import RenderContainer from '../RenderContainer';
import RenderLayer from '../RenderLayer';
import ZIndex from '../ZIndex';
import { Transition } from 'react-transition-group';
import raf from 'raf';

import PopupHelper, { Rect, PositionObject } from './PopupHelper';
import PopupPin from './PopupPin';
import LayoutEvents from '../../lib/LayoutEvents';

import styles from './Popup.less';

import { isIE } from '../ensureOldIEClassName';
import { Nullable } from '../../typings/utility-types';
import warning from 'warning';

const POPUP_BORDER_DEFAULT_COLOR = 'transparent';

export type PopupPosition =
  | 'top left'
  | 'top center'
  | 'top right'
  | 'right top'
  | 'right middle'
  | 'right bottom'
  | 'bottom left'
  | 'bottom center'
  | 'bottom right'
  | 'left top'
  | 'left middle'
  | 'left bottom';

export const PopupPositions: PopupPosition[] = [
  'top left',
  'top center',
  'top right',
  'right top',
  'right middle',
  'right bottom',
  'bottom right',
  'bottom center',
  'bottom left',
  'left bottom',
  'left middle',
  'left top'
];

export interface PopupHandlerProps {
  onMouseEnter?: (event: React.MouseEvent<HTMLElement> | MouseEvent) => void;
  onMouseLeave?: (event: React.MouseEvent<HTMLElement> | MouseEvent) => void;
  onClick?: (event: React.MouseEvent<HTMLElement> | MouseEvent) => void;
  onFocus?: (event: React.FocusEvent<HTMLElement> | FocusEvent) => void;
  onBlur?: (event: React.FocusEvent<HTMLElement> | FocusEvent) => void;
  onOpen?: () => void;
}

export interface PopupProps extends PopupHandlerProps {
  anchorElement: React.ReactNode | HTMLElement;
  backgroundColor?: React.CSSProperties['backgroundColor'];
  borderColor?: React.CSSProperties['borderColor'];
  children: React.ReactNode | (() => React.ReactNode);
  hasPin: boolean;
  hasShadow: boolean;
  disableAnimations: boolean;
  margin: number;
  maxWidth: number | string;
  opened: boolean;
  pinOffset: number;
  pinSize: number;
  popupOffset: number;
  positions: string[];
  onCloseRequest?: () => void;
}

interface PopupLocation {
  coordinates: {
    left: number;
    top: number;
  };
  position: string;
}

export interface PopupState {
  location: Nullable<PopupLocation>;
}

export default class Popup extends React.Component<PopupProps, PopupState> {
  public static propTypes = {
    /**
     * Ссылка (ref) на элемент или React компонент, для которого рисуется попап
     */
    anchorElement: PropTypes.oneOfType([
      PropTypes.instanceOf(HTMLElement),
      PropTypes.node
    ]).isRequired,

    /**
     * Фон попапа и пина
     */
    backgroundColor: PropTypes.string,

    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,

    /**
     * Показывать ли пин
     */
    hasPin: PropTypes.bool,

    /**
     * Применять ли box-shadow на попапе. При false отключает границу на пине
     */
    hasShadow: PropTypes.bool,

    /**
     * Отступ попапа от элемента
     */
    margin: PropTypes.number,

    /**
     * Показан или скрыт попап
     */
    opened: PropTypes.bool,

    /**
     * Смещение пина от края попапа. Край задаётся в пропе position вторым словом
     */
    pinOffset: PropTypes.number,

    /**
     * Сторона пина без учёта границы.
     * Пин представляет собой равносторонний треугольник, высота от попапа
     * до "носика" пина будет соответствовать формуле (size* √3)/2
     */
    pinSize: PropTypes.number,

    /**
     * смещение попапа относительно родительского элемента
     */
    popupOffset: PropTypes.number,

    /**
     * С какой стороны показывать попап и край попапа,
     * на котором будет отображаться пин
     */
    positions: PropTypes.array
  };

  public static defaultProps = {
    margin: 10,
    popupOffset: 0,
    pinSize: 8,
    pinOffset: 16,
    hasPin: false,
    hasShadow: false,
    disableAnimations: false,
    maxWidth: 500
  };

  public state: PopupState = {
    location: null
  };

  private layoutEventsToken: Nullable<
    ReturnType<typeof LayoutEvents.addListener>
  >;
  private lastPopupElement: Nullable<HTMLElement>;
  private anchorElement: Nullable<Element | Text>;
  private anchorInstance: Nullable<React.ReactInstance>;

  public componentDidMount() {
    this.updateLocation();
    this.layoutEventsToken = LayoutEvents.addListener(this.handleLayoutEvent);
  }

  public componentWillReceiveProps(nextProps: PopupProps) {
    /**
     * For react < 16 version ReactDOM.unstable_renderSubtreeIntoContainer is
     * used. It causes refs callbacks to call after componentDidUpdate.
     *
     * Delaying updateLocation to ensure that ref is set
     */
    if (!this.props.opened && nextProps.opened) {
      this.delayUpdateLocation();
    }

    if (this.props.opened && !nextProps.opened) {
      this.resetLocation();
    }
  }

  public componentDidUpdate(prevProps: PopupProps, prevState: PopupState) {
    if (
      prevState.location === null &&
      this.state.location !== null &&
      this.props.onOpen
    ) {
      this.props.onOpen();
    }
  }

  public componentWillUnmount() {
    if (this.layoutEventsToken) {
      this.layoutEventsToken.remove();
    }
  }

  public render() {
    const location = this.state.location || this.getDummyLocation();
    let child = null;

    if (this.props.anchorElement instanceof HTMLElement) {
      this.anchorElement = this.props.anchorElement;
    } else if (React.isValidElement(this.props.anchorElement)) {
      child = React.Children.only(this.props.anchorElement);
    } else {
      child = (
        <span
          onMouseEnter={this.handleMouseEnter}
          onMouseLeave={this.handleMouseLeave}
          onClick={this.handleClick}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        >
          {this.props.anchorElement}
        </span>
      );
    }

    return (
      <RenderLayer
        onClickOutside={this.handleClickOutside}
        onFocusOutside={this.handleFocusOutside}
        /**
         * If onCloseRequest is not specified handleClickOutside and handleFocusOutside
         * are doing nothing. So there is no need in RenderLayer at all.
         */
        active={Boolean(this.props.onCloseRequest) && this.props.opened}
      >
        <RenderContainer
          anchor={child}
          ref={child ? this.refAnchorElement : undefined}
        >
          {this.renderContent(location)}
        </RenderContainer>
      </RenderLayer>
    );
  }

  private updateAnchorElement(instance: React.ReactInstance) {
    const element = findDOMNode(instance);

    if (element === this.anchorElement) {
      return;
    }
    const { anchorElement } = this;
    if (anchorElement && anchorElement instanceof HTMLElement) {
      anchorElement.removeEventListener('mouseenter', this.handleMouseEnter);
      anchorElement.removeEventListener('mouseleave', this.handleMouseLeave);
      anchorElement.removeEventListener('click', this.handleClick);
      anchorElement.removeEventListener('focus', this.handleFocus);
      anchorElement.removeEventListener('blur', this.handleBlur);
    }

    this.anchorElement = element;

    if (element && element instanceof HTMLElement) {
      element.addEventListener('mouseenter', this.handleMouseEnter);
      element.addEventListener('mouseleave', this.handleMouseLeave);
      element.addEventListener('click', this.handleClick);
      element.addEventListener('focus', this.handleFocus);
      element.addEventListener('blur', this.handleBlur);
    }
  }

  private refAnchorElement = (instance: React.ReactInstance | null) => {
    this.anchorInstance = instance;
    if (this.anchorInstance) {
      this.updateAnchorElement(this.anchorInstance);
    } else {
      this.anchorElement = null;
    }
  };
  private handleMouseEnter = (
    event: React.MouseEvent<HTMLElement> | MouseEvent
  ) => {
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(event);
    }
  };
  private handleMouseLeave = (
    event: React.MouseEvent<HTMLElement> | MouseEvent
  ) => {
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(event);
    }
  };
  private handleClick = (event: React.MouseEvent<HTMLElement> | MouseEvent) => {
    if (this.props.onClick) {
      this.props.onClick(event);
    }
  };
  private handleFocus = (event: React.FocusEvent<HTMLElement> | FocusEvent) => {
    if (this.props.onFocus) {
      this.props.onFocus(event);
    }
  };
  private handleBlur = (event: React.FocusEvent<HTMLElement> | FocusEvent) => {
    if (this.props.onBlur) {
      this.props.onBlur(event);
    }
  };

  private renderContent(location: PopupLocation) {
    const { direction } = PopupHelper.getPositionObject(location.position);
    const { backgroundColor } = this.props;
    const rootStyle: React.CSSProperties = {
      top: location.coordinates.top,
      left: location.coordinates.left,
      maxWidth: this.props.maxWidth
    };

    return (
      <Transition
        timeout={{ enter: 0, exit: 200 }}
        appear={!this.props.disableAnimations}
        in={this.props.opened}
        mountOnEnter
        unmountOnExit
        enter={!this.props.disableAnimations}
        exit={!this.props.disableAnimations}
      >
        {(state: string) => (
          <ZIndex
            key={this.state.location ? 'real' : 'dummy'}
            delta={1000}
            ref={this.refPopupElement}
            className={cn({
              [styles.popup]: true,
              [styles.shadow]: this.props.hasShadow,
              [styles['transition-enter']]: state === 'entering',
              [styles['transition-enter-active']]: state === 'entered',
              [styles['transition-exit']]: state === 'exiting',
              [styles[
                ('transition-enter-' + direction) as keyof typeof styles
              ]]: true
            })}
            style={rootStyle}
            onMouseEnter={this.props.onMouseEnter}
            onMouseLeave={this.props.onMouseLeave}
          >
            <div className={styles.content}>
              <div className={styles.contentInner} style={{ backgroundColor }}>
                {this.renderChildren()}
              </div>
            </div>
            {this.renderPin(location.position)}
          </ZIndex>
        )}
      </Transition>
    );
  }

  private renderChildren() {
    return typeof this.props.children === 'function'
      ? this.props.children()
      : this.props.children;
  }

  private refPopupElement = (zIndex: ZIndex | null) => {
    if (zIndex) {
      this.lastPopupElement = zIndex && (findDOMNode(zIndex) as HTMLElement);
    }
  };

  private renderPin(position: string): React.ReactNode {
    /**
     * Box-shadow does not appear under the pin. Borders are used instead.
     * In non-ie browsers drop-shodow filter is used. It is applying
     * shadow to pin too.
     */
    const pinBorder =
      styles.popupBorderColor === POPUP_BORDER_DEFAULT_COLOR && isIE
        ? 'rgba(0, 0, 0, 0.09)'
        : styles.popupBorderColor;

    const {
      pinSize,
      pinOffset,
      hasShadow,
      backgroundColor,
      borderColor
    } = this.props;

    return (
      this.props.hasPin && (
        <PopupPin
          popupElement={this.lastPopupElement!}
          popupPosition={position}
          size={pinSize}
          offset={pinOffset}
          borderWidth={hasShadow ? 1 : 0}
          backgroundColor={backgroundColor || styles.popupBackground}
          borderColor={borderColor || pinBorder}
        />
      )
    );
  }

  private handleClickOutside = () => {
    this.requestClose();
  };

  private handleFocusOutside = () => {
    this.requestClose();
  };

  private handleLayoutEvent = () => {
    if (this.anchorInstance) {
      this.updateAnchorElement(this.anchorInstance);
    }
    if (this.state.location) {
      this.updateLocation();
    }
  };

  private delayUpdateLocation() {
    raf(() => this.updateLocation());
  }

  private updateLocation() {
    const popupElement = this.lastPopupElement;

    if (!popupElement) {
      return;
    }

    const location = this.getLocation(popupElement, this.state.location);
    if (!this.locationEquals(this.state.location, location)) {
      this.setState({ location });
    }
  }

  private resetLocation = () => {
    return this.setState({ location: null });
  };

  private requestClose = () => {
    if (this.props.onCloseRequest) {
      this.props.onCloseRequest();
    }
  };

  private locationEquals(
    x: Nullable<PopupLocation>,
    y: Nullable<PopupLocation>
  ) {
    if (x === y) {
      return true;
    }

    if (x == null || y == null) {
      return false;
    }

    return (
      x.coordinates.left === y.coordinates.left &&
      x.coordinates.top === y.coordinates.top &&
      x.position === y.position
    );
  }

  private getDummyLocation() {
    return {
      coordinates: {
        top: -9999,
        left: -9999
      },
      position: 'top left'
    };
  }

  private getLocation(
    popupElement: HTMLElement,
    location?: PopupLocation | null
  ) {
    const { positions, margin, popupOffset } = this.props;
    const { anchorElement } = this;

    if (process.env.NODE_ENV !== 'production') {
      warning(
        anchorElement && anchorElement instanceof HTMLElement,
        'Anchor element is not defined or not instance of HTMLElement'
      );
    }
    if (!(anchorElement && anchorElement instanceof HTMLElement)) {
      return location;
    }

    const anchorRect = PopupHelper.getElementAbsoluteRect(anchorElement);
    const popupRect = PopupHelper.getElementAbsoluteRect(popupElement);

    if (location && location.position) {
      // tslint:disable-next-line:no-shadowed-variable
      const position = PopupHelper.getPositionObject(location.position);
      // tslint:disable-next-line:no-shadowed-variable
      const coordinates = this.getCoordinates(
        anchorRect,
        popupRect,
        position,
        margin,
        popupOffset + this.getPinnedPopupOffset(anchorRect, position)
      );
      return { coordinates, position: location.position };
    }

    // tslint:disable-next-line:no-shadowed-variable
    for (const position of positions) {
      const positionObj = PopupHelper.getPositionObject(position);
      // tslint:disable-next-line:no-shadowed-variable
      const coordinates = this.getCoordinates(
        anchorRect,
        popupRect,
        positionObj,
        margin,
        popupOffset + this.getPinnedPopupOffset(anchorRect, positionObj)
      );
      if (
        PopupHelper.isAbsoluteRectFullyVisible({
          top: coordinates.top,
          left: coordinates.left,
          height: popupRect.height,
          width: popupRect.width
        })
      ) {
        return { coordinates, position };
      }
    }
    const position = PopupHelper.getPositionObject(positions[0]);
    const coordinates = this.getCoordinates(
      anchorRect,
      popupRect,
      position,
      margin,
      popupOffset + this.getPinnedPopupOffset(anchorRect, position)
    );
    return { coordinates, position: positions[0] };
  }

  private getPinnedPopupOffset(anchorRect: Rect, position: PositionObject) {
    if (!this.props.hasPin || /center|middle/.test(position.align)) {
      return 0;
    }

    const anchorSize = /top|bottom/.test(position.direction)
      ? anchorRect.width
      : anchorRect.height;

    const { pinOffset, pinSize } = this.props;
    return Math.max(0, pinOffset + pinSize - anchorSize / 2);
  }

  private getCoordinates(
    anchorRect: Rect,
    popupRect: Rect,
    position: PositionObject,
    margin: number,
    popupOffset: number
  ) {
    switch (position.direction) {
      case 'top':
        return {
          top: anchorRect.top - popupRect.height - margin,
          left: this.getHorizontalPosition(
            anchorRect,
            popupRect,
            position.align,
            popupOffset
          )
        };
      case 'bottom':
        return {
          top: anchorRect.top + anchorRect.height + margin,
          left: this.getHorizontalPosition(
            anchorRect,
            popupRect,
            position.align,
            popupOffset
          )
        };
      case 'left':
        return {
          top: this.getVerticalPosition(
            anchorRect,
            popupRect,
            position.align,
            popupOffset
          ),
          left: anchorRect.left - popupRect.width - margin
        };
      case 'right':
        return {
          top: this.getVerticalPosition(
            anchorRect,
            popupRect,
            position.align,
            popupOffset
          ),
          left: anchorRect.left + anchorRect.width + margin
        };
      default:
        throw new Error(`Unxpected direction '${position.direction}'`);
    }
  }

  private getHorizontalPosition(
    anchorRect: Rect,
    popupRect: Rect,
    align: string,
    popupOffset: number
  ) {
    switch (align) {
      case 'left':
        return anchorRect.left - popupOffset;
      case 'center':
        return anchorRect.left - (popupRect.width - anchorRect.width) / 2;
      case 'right':
        return (
          anchorRect.left - (popupRect.width - anchorRect.width) + popupOffset
        );
      default:
        throw new Error(`Unxpected align '${align}'`);
    }
  }

  private getVerticalPosition(
    anchorRect: Rect,
    popupRect: Rect,
    align: string,
    popupOffset: number
  ) {
    switch (align) {
      case 'top':
        return anchorRect.top - popupOffset;
      case 'middle':
        return anchorRect.top - (popupRect.height - anchorRect.height) / 2;
      case 'bottom':
        return (
          anchorRect.top - (popupRect.height - anchorRect.height) + popupOffset
        );
      default:
        throw new Error(`Unxpected align '${align}'`);
    }
  }
}
