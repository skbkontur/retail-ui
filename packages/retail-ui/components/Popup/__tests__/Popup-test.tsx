import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import Popup, { PopupProps, PopupState } from '../Popup';
import toJson from 'enzyme-to-json';
import { delay } from '../../../lib/utils';

const openPopup = async (wrapper: ReactWrapper<PopupProps, PopupState, Popup>) =>
  new Promise(async resolve => {
    wrapper.setProps({ opened: true }, async () => {
      await delay(100);
      resolve();
    });
  });

const closePopup = async (wrapper: ReactWrapper<PopupProps, PopupState, Popup>) =>
  new Promise(async resolve => {
    wrapper.setProps({ opened: false }, async () => {
      await delay(100);
      resolve();
    });
  });

const renderWrapper = (props?: Partial<PopupProps>): ReactWrapper<PopupProps, PopupState, Popup> => {
  const anchor = document.createElement('button');

  anchor.id = 'test-id';
  anchor.innerHTML = 'test';

  return mount(
    <Popup
      positions={['bottom left', 'bottom right', 'top left', 'top right']}
      opened={false}
      anchorElement={anchor as HTMLElement}
      {...props}
    >
      Test content
    </Popup>,
  );
};

describe('Popup', () => {
  jest.setTimeout(10000);

  it('по дефолту открывается в первой переданной позиции', async () => {
    const anchor = document.createElement('button');

    anchor.id = 'test-id';
    anchor.innerHTML = 'test';

    const wrapper = mount<Popup>((
      <Popup
        positions={['bottom right', 'top left', 'top right', 'bottom left']}
        opened={false}
        anchorElement={anchor as HTMLElement}
      >
        Test content
      </Popup>
    ) as React.ReactElement<PopupProps>);

    expect(wrapper.state('location')).toBeNull();

    await openPopup(wrapper);

    expect(wrapper.state('location')).not.toBeNull();
    expect(wrapper.state('location')!.position).toBe('bottom right');
  });

  it('одна и та же позиция при каждом открытии', async () => {
    const anchor = document.createElement('button');

    anchor.id = 'test-id';
    anchor.innerHTML = 'test';
    anchor.style.setProperty('margin', '0 0 100px 0');

    const wrapper = mount<Popup>((
      <Popup
        positions={['top left', 'top right', 'bottom left', 'bottom right']}
        opened={false}
        anchorElement={anchor as HTMLElement}
        popupOffset={20}
      >
        Test content
      </Popup>
    ) as React.ReactElement<PopupProps>);

    expect(wrapper.state('location')).toBeNull();

    const check = () =>
      new Promise(async resolve => {
        await openPopup(wrapper);

        expect(wrapper.state('location')!.position).toBe('bottom right');

        await closePopup(wrapper);

        expect(wrapper.state('location')).toBeNull();

        resolve();
      });

    const checkLocation = async () => {
      for (let i = 0; i < 11; i++) {
        await check();
      }
    };

    await checkLocation();
  });
});

describe('Popup snapshots: ', () => {
  it('open and close popup', async () => {
    const wrapper = renderWrapper();

    expect(toJson(wrapper)).toMatchSnapshot('closed popup');

    await openPopup(wrapper);
    expect(toJson(wrapper)).toMatchSnapshot('opened popup');

    await closePopup(wrapper);
    wrapper.update();
    /** Ждем когда кончится анимация react-transition-group и случится unmount */
    await delay(300);
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot('closed popup again');
  });
});
