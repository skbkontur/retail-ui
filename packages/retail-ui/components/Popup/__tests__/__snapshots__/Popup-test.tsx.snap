// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup snapshots:  open and close popup: closed popup 1`] = `
<Popup
  anchorElement={
    <button
      id="test-id"
    >
      test
    </button>
  }
  backgroundColor="#fff"
  disableAnimations={false}
  hasPin={false}
  hasShadow={false}
  margin={10}
  maxWidth={500}
  opened={false}
  pinOffset={16}
  pinSize={8}
  popupOffset={0}
  positions={
    Array [
      "bottom left",
      "bottom right",
      "top left",
      "top right",
    ]
  }
>
  <RenderLayer
    active={true}
    onClickOutside={[Function]}
    onFocusOutside={[Function]}
  >
    <RenderContainer>
      <Transition
        appear={true}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={true}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={
          Object {
            "enter": 0,
            "exit": 200,
          }
        }
        unmountOnExit={true}
      />
      <Portal
        key="portal-ref"
        rt_rootID={3}
      >
        <noscript
          data-render-container-id={3}
        />
      </Portal>
    </RenderContainer>
  </RenderLayer>
</Popup>
`;

exports[`Popup snapshots:  open and close popup: closed popup again 1`] = `
<Popup
  anchorElement={
    <button
      id="test-id"
    >
      test
    </button>
  }
  backgroundColor="#fff"
  disableAnimations={false}
  hasPin={false}
  hasShadow={false}
  margin={10}
  maxWidth={500}
  opened={false}
  pinOffset={16}
  pinSize={8}
  popupOffset={0}
  positions={
    Array [
      "bottom left",
      "bottom right",
      "top left",
      "top right",
    ]
  }
>
  <RenderLayer
    active={true}
    onClickOutside={[Function]}
    onFocusOutside={[Function]}
  >
    <RenderContainer>
      <Transition
        appear={true}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={true}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={
          Object {
            "enter": 0,
            "exit": 200,
          }
        }
        unmountOnExit={true}
      />
      <Portal
        key="portal-ref"
        rt_rootID={3}
      >
        <noscript
          data-render-container-id={3}
        />
      </Portal>
    </RenderContainer>
  </RenderLayer>
</Popup>
`;

exports[`Popup snapshots:  open and close popup: opened popup 1`] = `
<Popup
  anchorElement={
    <button
      id="test-id"
    >
      test
    </button>
  }
  backgroundColor="#fff"
  disableAnimations={false}
  hasPin={false}
  hasShadow={false}
  margin={10}
  maxWidth={500}
  opened={true}
  pinOffset={16}
  pinSize={8}
  popupOffset={0}
  positions={
    Array [
      "bottom left",
      "bottom right",
      "top left",
      "top right",
    ]
  }
>
  <RenderLayer
    active={true}
    onClickOutside={[Function]}
    onFocusOutside={[Function]}
  >
    <RenderContainer>
      <Transition
        appear={true}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={true}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={
          Object {
            "enter": 0,
            "exit": 200,
          }
        }
        unmountOnExit={true}
      >
        <ZIndex
          className="popup transition-enter-top"
          delta={1000}
          key="dummy"
          render={true}
          style={
            Object {
              "backgroundColor": "#fff",
              "left": -9999,
              "maxWidth": 500,
              "top": -9999,
            }
          }
        >
          <div
            className="popup transition-enter-top"
            style={
              Object {
                "backgroundColor": "#fff",
                "left": -9999,
                "maxWidth": 500,
                "top": -9999,
                "zIndex": 26000,
              }
            }
          >
            Test content
          </div>
        </ZIndex>
      </Transition>
      <Portal
        key="portal-ref"
        rt_rootID={3}
      >
        <noscript
          data-render-container-id={3}
        />
      </Portal>
    </RenderContainer>
  </RenderLayer>
</Popup>
`;
